README.md
[<img src="https://github-ads.s3.eu-central-1.amazonaws.com/support-ukraine.svg?t=1" />](https://supportukrainenow.org)

# Convert BMI values to other formats

[![Latest Version on Packagist](https://img.shields.io/packagist/v/jeroenboumans/php-bmi.svg?style=flat-square)](https://packagist.org/packages/jeroenboumans/php-bmi)
[![Tests](https://github.com/jeroenboumans/php-bmi/actions/workflows/run-tests.yml/badge.svg?branch=main)](https://github.com/jeroenboumans/php-bmi/actions/workflows/run-tests.yml)
[![Total Downloads](https://img.shields.io/packagist/dt/jeroenboumans/php-bmi.svg?style=flat-square)](https://packagist.org/packages/jeroenboumans/php-bmi)

This is where your description should go. Try and limit it to a paragraph or two. Consider adding a small example.

## Support us

[<img src="https://github-ads.s3.eu-central-1.amazonaws.com/BMI.jpg?t=1" width="419px" />](https://spatie.be/github-ad-click/BMI)

We invest a lot of resources into creating [best in class open source packages](https://spatie.be/open-source). You can support us by [buying one of our paid products](https://spatie.be/open-source/support-us).

We highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using. You'll find our address on [our contact page](https://spatie.be/about-us). We publish all received postcards on [our virtual postcard wall](https://spatie.be/open-source/postcards).

## Installation

You can install the package via composer:

```bash
composer require jeroenboumans/php-bmi
```

## Usage

```php
$skeleton = new JeroenBoumans\BMI();
echo $skeleton->echoPhrase('Hello, JeroenBoumans!');
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/spatie/.github/blob/main/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Jeroen Boumans](https://github.com/jeroenboumans)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `BMI` will be documented in this file.



tests/CalculateBMITest.php
<?php

use JeroenBoumans\BMI\BMI;
use JeroenBoumans\BMI\Enum\Gender;

test('BMI value can be calculated correctly', function () {
    expect(BMI::from(65, 170, 12, Gender::MALE)->calculate())->toEqual(22.5);
    expect(BMI::from(65, 170, 12, Gender::MALE)->calculate(2))->toEqual(22.49);
});


tests/ConvertTrafficLightTest.php
<?php

use JeroenBoumans\BMI\BMI;
use JeroenBoumans\BMI\Converter\TrafficLightConverter;
use JeroenBoumans\BMI\Enum\Gender;

test('Correct column chosen', function () {
    expect((new TrafficLightConverter())
        ->getCategoryValues(gender: Gender::MALE, age: 12))
        ->toEqual([13.18, 14.05, 21.22, 26.02]);
});

test('BMI value can be calculated correctly', function () {
    expect(BMI::from(weight: 65, length: 170, age: 18, gender: Gender::MALE)->toTrafficLight())->toEqual('orange');
});


src/BMI.php
<?php

namespace JeroenBoumans\BMI;

use JeroenBoumans\BMI\Converter\TrafficLightConverter;
use JeroenBoumans\BMI\Enum\Gender;

class BMI
{
    /**
     * @param float $weight in kg
     * @param float $length in cm
     * @return void
     */
    public function __construct(
        public readonly float $weight,
        public readonly float $length,
        public readonly float $age,
        public readonly Gender $gender
    ) {
    }

    /**
     * @param float $weight in kg
     * @param float $length in cm
     * @param float $age
     * @param Gender $gender
     * @return self
     */
    public static function from(float $weight, float $length, float $age, Gender $gender): self
    {
        return new self(
            weight: $weight,
            length: $length,
            age: $age,
            gender: $gender
        );
    }

    /**
     * @return string
     */
    public function toTrafficLight(): string
    {
        return TrafficLightConverter::lookUp(
            gender: $this->gender,
            age: $this->age,
            bmi: $this->calculate()
        );
    }

    /**
     * @param int $precision
     * @return float
     */
    public function calculate(int $precision = 1): float
    {
        return round($this->weight / pow($this->length / 100, 2), $precision);
    }
}


src/Converter/TrafficLightConverter.php
<?php

namespace JeroenBoumans\BMI\Converter;

use JeroenBoumans\BMI\Enum\Category;
use JeroenBoumans\BMI\Enum\Gender;

class TrafficLightConverter
{
    public static function lookUp(Gender $gender, int $age, float $bmi): string
    {
        $instance = new self();
        $columnValues = $instance->getCategoryValues(gender: $gender, age: $age);
        $category = null;

        foreach ($columnValues as $index => $value) {
            if ($bmi <= $value) {
                $category = self::categories[$index];
            }
        }

        return match ($category) {
            default => "red",
            Category::OVERWEIGHT,
            Category::UNDERWEIGHT => "orange",
            Category::NORMAL => "green"
        };
    }

    public function getCategoryValues(Gender $gender, int $age): array
    {
        if ($age > 1 && $age <= 18) {
            return self::values[ match ($gender) {
                Gender::FEMALE => "f",
                default => "m"
            }][$age];
        }

        return self::values["adult"];
    }

    private const categories = [
        Category::EXTREMELY_UNDERWEIGHT,
        Category::UNDERWEIGHT,
        Category::NORMAL,
        Category::OVERWEIGHT,

        // when none of the above apply
        Category::OBESE,
    ];

    private const values = [
        "adult" => [ 18.5, 25.0, 30.0, 40.0 ],
        "f" => [
            2 => [13.24, 13.9, 18.02, 19.81],
            3 => [12.98, 13.6, 17.56, 19.36],
            4 => [12.73, 13.34, 17.28, 19.15],
            5 => [12.5, 13.09, 17.15, 19.17],
            6 => [12.32, 12.93, 17.34, 19.65],
            7 => [12.26, 12.91, 17.75, 20.51],
            8 => [12.31, 13, 18.35, 21.57],
            9 => [12.44, 13.18, 19.07, 22.81],
            10 => [12.64, 13.43, 19.86, 24.11],
            11 => [12.95, 13.79, 20.74, 25.42],
            12 => [13.39, 14.28, 21.68, 26.67],
            13 => [13.92, 14.85, 22.58, 27.76],
            14 => [14.48, 15.43, 23.34, 28.57],
            15 => [15.01, 15.98, 23.94, 29.11],
            16 => [15.46, 16.44, 24.37, 29.43],
            17 => [15.78, 16.77, 24.7, 29.69],
            18 => [16, 17, 25, 30],
        ],
        "m" => [
            2 => [13.37, 14.12, 18.41, 20.09],
            3 => [13.09, 13.79, 17.89, 19.57],
            4 => [12.86, 13.52, 17.55, 19.29],
            5 => [12.66, 13.31, 17.42, 19.3],
            6 => [12.5, 13.15, 17.55, 19.78],
            7 => [12.42, 13.08, 17.92, 20.63],
            8 => [12.42, 13.11, 18.44, 21.6],
            9 => [12.5, 13.24, 19.1, 22.77],
            10 => [12.66, 13.45, 19.84, 24],
            11 => [12.89, 13.72, 20.55, 25.1],
            12 => [13.18, 14.05, 21.22, 26.02],
            13 => [13.59, 14.48, 21.91, 26.84],
            14 => [14.09, 15.01, 22.62, 27.63],
            15 => [14.6, 15.55, 23.29, 28.3],
            16 => [15.12, 16.08, 23.9, 28.88],
            17 => [15.6, 16.58, 24.46, 29.41],
            18 => [16, 17, 25, 30],
        ],
    ];
}
